File: plugin/src/main/com/suhininalex/clones/CloneIterator.java
//package com.suhininalex.clones;
//
//import com.suhininalex.suffixtree.Edge;
//import com.suhininalex.suffixtree.Node;
//
//import java.util.Iterator;
//import java.util.List;
//import java.util.Stack;
//
///**
// * Pre-order traversal
// */
//class CloneIterator implements Iterator<Clone> {
//
//    private final Stack<Edge> stack = new Stack<>();
//    private final int cloneLength;
//
//    private void addNodeToStack(Node node){
//        stack.addAll(node.getEdges());
//    }
//
//    public CloneIterator(Node node, int cloneLength){
//        this.cloneLength = cloneLength;
//        if (cloneLength>=0) addNodeToStack(node);
//    }
//
//    @Override
//    public boolean hasNext() {
//        return !stack.isEmpty();
//    }
//
//    @Override
//    public Clone next() {
//        Edge edge = stack.pop();
//
//        if (edge.getTerminal()!=null) {
//            addNodeToStack(edge.getTerminal());
//            return next();
//        }
//
//        int lastElementIndex = edge.getParent().getParentEdge().getEnd();
//        List<Token> sequence = edge.getSequence();
//        return new Clone(sequence.get(lastElementIndex - cloneLength + 1), sequence.get(lastElementIndex));
//    }
//}

File: plugin/src/main/com/suhininalex/clones/CloneIterator.kt
package com.suhininalex.clones

import com.suhininalex.suffixtree.Edge
import com.suhininalex.suffixtree.Node
import java.util.*

class CloneIterator(val cloneClass: CloneClass) : Iterator<Clone> {

    private val stack = Stack<EdgeInfo>()

    private fun addEdgesToStack(node : Node, offset: Int) =
        node.edges.forEach {
            stack.push(EdgeInfo(it, offset))
        }

    init {
        addEdgesToStack(cloneClass.treeNode, 0)
    }

    override fun hasNext(): Boolean {
        return !stack.isEmpty()
    }

    override fun next(): Clone {
        val( edge, offset) = stack.pop()

        with (edge) {
            if (terminal != null) {
                addEdgesToStack(edge.terminal, offset+edge.length)
                return next()
            } else {
                val lastElementIndex = end-offset-length
                val firstElementIndex = lastElementIndex - cloneClass.length + 1
                return Clone(getFromSequence(firstElementIndex),getFromSequence(lastElementIndex))
            }
        }
    }

    private data class EdgeInfo(val edge: Edge, val offset: Int)

    private fun Edge.getFromSequence(pos: Int) = sequence[pos] as Token
}